cpp/src/arrow/util/reflection_test.cc

```diff
@@ -0,0 +1,121 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <gtest/gtest.h>
+
+#include "arrow/util/reflection_internal.h"
+
+namespace arrow {
+namespace internal {
+
+// unmodified structure which we wish to reflect on:
+struct Person {
+  std::string name;
+  int age;
+};
+
+// enumeration of properties:
+template <>
+struct ReflectionTraits<Person> {
+  using Properties = std::tuple<DataMember<Person, int, &Person::age>,
```

Comment:
Is reversing the order intentional? It confused me for a bit below when I saw that the properties were in the opposite order of the struct until I looked here again.

Reply:
It was intentional; I wanted to make clear that the ordering follows the trait and not the declaration in the struct. But in practice we'd probably always order identically. I'll switch them

Reply:
Ah that's fine too - might just deserve a comment so it's clear what the intent was.